packages/paste-core/components/code-block/__tests__/index.spec.tsx

```diff
@@ -0,0 +1,336 @@
+          <CodeBlock language="ruby" data-testid="ruby-block">{`#!/usr/bin/ruby`}</CodeBlock>;
+        </CodeBlockGroup>
+      </Theme.Provider>
+    );
+
+    const tablist = screen.getByRole('tablist');
+    expect(tablist).toBeDefined();
+
+    const javascriptTab = screen.getByRole('tab', {name: 'JavaScript'});
+    expect(javascriptTab).toBeDefined();
+    expect(javascriptTab.getAttribute('aria-selected')).toBe('true');
+
+    const javascriptTabPanel = screen.getByTestId('js-block').parentElement;
+    expect(javascriptTabPanel).toBeDefined();
+    expect(javascriptTab.getAttribute('aria-controls')).toBe(javascriptTabPanel?.id);
+    expect(javascriptTabPanel?.getAttribute('aria-labelledby')).toBe(javascriptTab.id);
+
+    const rubyTab = screen.getByRole('tab', {name: 'Ruby'});
+    expect(rubyTab).toBeDefined();
+    expect(rubyTab.getAttribute('aria-selected')).toBe('false');
+
+    const rubyTabPanel = screen.getByTestId('ruby-block').parentElement;
+    expect(rubyTabPanel).toBeDefined();
+    expect(rubyTab.getAttribute('aria-controls')).toBe(rubyTabPanel?.id);
+    expect(rubyTabPanel?.getAttribute('aria-labelledby')).toBe(rubyTab.id);
+  });
+
+  describe('i18n', () => {
+    it('should pass i18n props to CopyButton and ExternalLinkButton', () => {
+      render(
+        <Theme.Provider theme="default">
+          <CodeBlockGroup
+            label="My code block group"
+            i18nCopyTooltipTextAfter="after text"
+            i18nCopyTooltipTextBefore="before text"
+            i18nLinkDescription="link description"
+          >
+            <CodeBlock language="javascript" externalLink="https://www.google.com">{`(num) => num + 1`}</CodeBlock>;
+            <CodeBlock language="ruby">{`#!/usr/bin/ruby`}</CodeBlock>;
+          </CodeBlockGroup>
+        </Theme.Provider>
+      );
+
+      const beforeText = screen.getAllByText('before text');
+      // it appears twice in the dom, once in the button text and once in the tooltip
+      expect(beforeText).toHaveLength(2);
+
+      // TODO: make this test work
+      // const afterText = screen.getAllByText('after text');
+      // expect(afterText).toHaveLength(2)
+
+      const linkDescription = screen.getAllByText('link description');
+      // it appears twice in the dom, once in the button text and once in the tooltip
+      expect(linkDescription).toHaveLength(2);
+    });
+  });
+});
+
+describe('CopyButton', () => {
+  it('should render', () => {
+    render(
+      <Theme.Provider theme="default">
+        <CopyButton text="hello world" />
+      </Theme.Provider>
+    );
+
+    const button = screen.getByRole('button');
+    const tooltip = screen.getByRole('tooltip', {hidden: true});
+
+    expect(button).toBeDefined();
+    expect(tooltip).not.toBeVisible();
+  });
+
+  describe('i18n', () => {
+    //   Note: if click copy button, need this to mock document.commandExec, otherwise throws error
```

Comment:
Current issue: For some reason, I'm having a ton of trouble testing that the Copy code block button text changes to "Copied!" after you click it. I currently commented these out, but I need the mock of `document.commandExec` because the `useClipboardCopy` hook calls it (this code is taken from use-clipboard-copy's test files: https://github.com/wsmd/use-clipboard-copy/blob/master/test/index.test.tsx).

In the test "should have default copy before and after text" below, I have tried:
* wrapping the fireEvent in waitFor and act to make sure the DOM is fully updated before I look for the new string
* using `jest.useFakeTimers` to take a little extra time after click

I know that it should be working because when I look at the button in storybook, it behaves as expected.

I don't know whats going on, but it would be really nice if I could test that this functionality works 🤷🏽 .

Reply:
Solution: we simplified the tests! instead of checking the copy behavior, added a new function that just returns the correct string. There is a test for just that function that makes sure it returns the correct string based on the copied arg.